### PR TITLE
[v3] Fix and optimise assetserver

### PR DESCRIPTION
**Pull request by fbbdev**
# Description

This PR fixes a couple issues in the assetserver, optimises response wrappers and provides some minor code cleanup.

Fixed issues:

- Non-404 responses on the root index path (`/`, `/index`, `/index.html`) were not being handled correctly (the status code was being forwarded, but headers and body were being discarded).
- The content-type sniffer tried to sniff MIME types from data passed in the first call to Write, without providing the full 512 bytes that might be required. Now an initial prefix of at most 512 bytes is cached and passed to the sniffing method.

Optimisations:

- The fallback system for the root index path has been made zero-copy, i.e. completely avoids buffering the nested response.
- The content-type sniffer caches data in a conservative way, avoiding copies at all whenever possible.

Cleanup:

- Fixed some minor warnings from the compiler/static checker.
- Removed useless calls to formatting functions.
- Fixed logging calls that treated logging methods as formatters, when they are not.

## Type of change
  
Please select the option that is relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Ran unit tests and examples, tested various edge cases by hand, added new tests for content sniffing and index fallback.
  
- [ ] Windows
- [x] macOS
- [ ] Linux

## Test Configuration

```
 Wails (v3.0.0-dev)  Wails Doctor 
                                                                                                                                                    
# System 

┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
| Name          | MacOS                                                                                                          |
| Version       | 12.6.6                                                                                                         |
| ID            | 21G646                                                                                                         |
| Branding      | Monterey                                                                                                       |
| Platform      | darwin                                                                                                         |
| Architecture  | amd64                                                                                                          |
| Apple Silicon | unknown                                                                                                        |
| CPU           | Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz                                                                       |
| CPU           | Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz                                                                       |
| GPU           |  cores, Metal Family: Supported, Metal GPUFamily macOS 2                                                       |
|       Metal Family: Supported, Metal GPUFamily macOS 2                                                                         |
| Memory        | 32 GB                                                                                                          |
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

# Build Environment 

┌─────────────────────────────────────────────────────────┐
| Wails CLI    | v3.0.0-dev                               |
| Go Version   | go1.23.5                                 |
| Revision     | b88d57cbd960c985bd57bab3e13ead10d1618da9 |
| Modified     | false                                    |
| -buildmode   | exe                                      |
| -compiler    | gc                                       |
| CGO_CFLAGS   |                                          |
| CGO_CPPFLAGS |                                          |
| CGO_CXXFLAGS |                                          |
| CGO_ENABLED  | 1                                        |
| CGO_LDFLAGS  |                                          |
| GOAMD64      | v1                                       |
| GOARCH       | amd64                                    |
| GOOS         | darwin                                   |
| vcs          | git                                      |
| vcs.modified | false                                    |
| vcs.revision | b88d57cbd960c985bd57bab3e13ead10d1618da9 |
| vcs.time     | 2025-02-06T14:05:25Z                     |
└─────────────────────────────────────────────────────────┘

# Dependencies 

┌───────────────────────────┐
| Xcode cli tools | 2395    |
| npm             | 10.8.2  |
| *NSIS           | v3.10   |
└─ * - Optional Dependency ─┘

# Checking for issues 

 SUCCESS  No issues found

# Diagnosis 

 SUCCESS  Your system is ready for Wails development!
```

# Checklist:

- [x] I have updated `website/src/pages/changelog.mdx` with details of this PR
- [x] My code follows the general coding style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

- **Documentation**
  - Updated the changelog to include recent bug fixes related to asset server response handling.

- **Bug Fixes**
  - Fixed size of response prefix used for content type sniffing in the asset server.
  - Resolved issues with handling non-404 responses on the root index path of the asset server.
  - Enhanced error handling and response management capabilities of the asset server.

- **New Features**
  - Introduced a fallback mechanism for handling 404 responses, ensuring default content is delivered when necessary.

- **Tests**
  - Added new test cases for content sniffing and index fallback functionalities to improve test coverage.

### CODE REVIEW

File: v3/internal/assetserver/content_type_sniffer.go [hunk 1]
```diff
@@ -5,38 +5,106 @@ import (
...
+		return 0, nil
+	}
+
+	// Cut away at most 512 bytes from chunk.
+	cut := min(len(chunk), 512-len(rw.prefix))
+	if cut >= 512 {
+		// Avoid copying data if a full prefix is available on first non-zero write.
+		cut = len(chunk)
+		rw.prefix = chunk
+		chunk = nil
```
[the remark this answers is not in the document]
Comment (fbbdev) [reply v3/internal/assetserver/content_type_sniffer.go:43]: @coderabbitai you're right that it's not immediately evident from the context whether this operation is safe or not. The underlying assumptions are:

- the chunk is only stored directly when it is 512 bytes or more in length;
- in that case, `rw.complete` is called immediately, which is guaranteed to reset `rw.prefix` to `nil`

in other words, the chunk is only referenced temporarily and there is no risk of modifying it.

I agree that, given the high level of non-contextuality, it would be easy to introduce changes that accidentally break the assumptions above. However:

- this piece of code is unlikely to change in the future;
- it is a necessary optimization to avoid copying a large slice, which is going to be _the most common case._

File: v3/internal/assetserver/assetserver_test.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+				if res.Code != status {
+					t.Errorf("Status code mismatch: want %d, got %d", status, res.Code)
+				}
+
+				if fallbackTriggered {
+					if cmp.Equal(body, res.Body.String()) {
+						t.Errorf("Fallback response has the same body as not found response")
+					}
+					return
+				} else {
```
[the remark this answers is not in the document]
Comment (fbbdev) [reply v3/internal/assetserver/assetserver_test.go:230]: @coderabbitai it would be possible to add such checks but they would rely upon knowledge of the internals of the assetserver, making it easy to break the test with benign changes. The exact content of the fallback response is not very important.

File: v3/internal/assetserver/content_type_sniffer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (fbbdev) [reply v3/internal/assetserver/content_type_sniffer.go]: On second thought, this is even safer than I stated, because the only place where `rw.prefix` is written is the other branch of the surrounding if statement, and it should never modify the slice when it has already 512 bytes or more.

I added a couple safety checks. They are not necessary, but they ensure the behaviour of this fragment will be locally correct independently of the implementation details of other methods.

File: v3/internal/assetserver/webview/responsewriter_windows.go [hunk 3]
```diff
@@ -68,7 +69,7 @@ func (rw *responseWriter) Finish() error {
 	if code == http.StatusNotModified {
 		// WebView2 has problems when a request returns a 304 status code and the WebView2 is going to hang for other
```
Comment (leaanthony) [inline v3/internal/assetserver/webview/responsewriter_windows.go:70]: This was committed in September 2023. I've scoured the Webview2 feedback site and can't find anything about this status code 🤔

**Review by leaanthony — APPROVED** [review]

**Review by leaanthony — APPROVED** [review]
